client/grpc/client.go

```diff
@@ -0,0 +1,164 @@
+/*
+ * Flow Go SDK
+ *
+ * Copyright 2019 Dapper Labs, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+// Package grpc client provides a Go client for the Flow Access gRPC API.
+//
+// The Access API provides a set of methods that can be used to submit transactions
+// and read state from Flow. This client is compatible with the Access API implemented by the
+// Access Node role, as well as the mock Access API exposed by the Flow Emulator.
+//
+// The full Access API specification is here: https://docs.onflow.org/access-api/
+package grpc
+
+import (
+	"context"
+
+	"google.golang.org/grpc"
+
+	"github.com/onflow/cadence"
+	"github.com/onflow/flow-go-sdk"
+	"google.golang.org/grpc/credentials/insecure"
+)
+
+const EMULATOR_URL = "127.0.0.1:3569"
+const TESTNET_URL = "access.devnet.nodes.onflow.org:9000"
+const CANARYNET_URL = "access.canary.nodes.onflow.org:9000"
+const MAINNET_URL = "access.mainnet.nodes.onflow.org:9000"
+
+// NewClient create a client by passing the gRPC handler.
+func NewClient(url string) (*BaseClient, error) {
+	client, err := NewGRPCClient(url, grpc.WithTransportCredentials(insecure.NewCredentials()))
+	if err != nil {
+		return nil, err
+	}
+
+	return &BaseClient{client}, nil
+}
+
+// BaseClient complies with the client interface and hides any gRPC specific options.
+type BaseClient struct {
```

Comment:
Hmm, I think I would actually flip these names around.

- ~`BaseClient`~ becomes `Client` and is the struct that conforms to `access.Client` and is interchangeable with `http.Client`.
- ~`GRPCClient`~ becomes `BaseClient` and is the underlying struct that provides the base functionality for `Client` and also exposes more advanced features. This also avoids repetition with the package name (`grpc.GRPCClient` repeats GRPC twice).

You then get code like this:

```go
// this feels like the sensible default
client, err := grpc.NewClient()

// this feels like more advanced behaviour
baseClient, err := grpc.NewBaseClient()

Reply:
I agree. I actually didn't understand our conversation correctly and hence made this mess. This way it's way more clear.